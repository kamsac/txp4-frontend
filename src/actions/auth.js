import jwtDecode from 'jwt-decode';
import dotenvConfiguration from '../dotenv-configuration';
import AuthResource from '../resources/auth';
import AuthResourceMock from '../resources/auth/mock';

export const LOGIN_REQUEST = 'LOGIN_REQUEST';
export const LOGIN_SUCCESS = 'LOGIN_SUCCESS';
export const LOGIN_FAILURE = 'LOGIN_FAILURE';
export const LOGOUT_REQUEST = 'LOGOUT_REQUEST';
export const LOGOUT_SUCCESS = 'LOGOUT_SUCCESS';

export function requestLogin() {
  return {
    type: LOGIN_REQUEST,
    payload: {
      isFetching: true,
    },
  };
}

export function receiveLogin(accessToken) {
  return {
    type: LOGIN_SUCCESS,
    payload: {
      isFetching: false,
      jwtDecoded: jwtDecode(accessToken),
    },
  };
}

export function loginError() {
  return {
    type: LOGIN_FAILURE,
    payload: {
      isFetching: false,
      jwtDecoded: null,
    },
  };
}

export function loginUser(magicLinkToken) {
  const resource = dotenvConfiguration.API_URL ? AuthResource : AuthResourceMock;

  return (dispatch) => {
    dispatch(requestLogin());

    resource.login(magicLinkToken)
      .then((response) => {
        const { token } = response.data;
        localStorage.setItem('accessToken', token);
        dispatch(receiveLogin(token));
      })
      .catch(() => {
        dispatch(loginError());
      });
  };
}

export function requestLogout() {
  return {
    type: LOGOUT_REQUEST,
    payload: {},
  };
}

export function receiveLogout() {
  return {
    type: LOGOUT_SUCCESS,
    payload: {
      jwtDecoded: null,
    },
  };
}

export function logoutUser() {
  return (dispatch) => {
    dispatch(requestLogout());

    localStorage.removeItem('accessToken');

    dispatch(receiveLogout());
  };
}
