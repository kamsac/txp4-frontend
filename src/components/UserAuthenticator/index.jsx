import React from 'react';
import { Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';

export default class UserAuthenticator extends React.Component {
  componentDidMount() {
    this.props.loginUser(this.props.magicLinkToken);
  }

  render() {
    const {
      isUserAuthenticated,
      isFetching,
    } = this.props;

    if (isUserAuthenticated) {
      return (
        <Redirect to="/inventory" />
      );
    }

    if (!isUserAuthenticated && isFetching) {
      return <p>You are being authenticated...</p>;
    }

    return <p>Something went wrong while trying to authenticate.</p>;
  }
}

UserAuthenticator.propTypes = {
  magicLinkToken: PropTypes.string.isRequired,
  loginUser: PropTypes.func.isRequired,
  isUserAuthenticated: PropTypes.bool.isRequired,
  isFetching: PropTypes.bool.isRequired,
};
