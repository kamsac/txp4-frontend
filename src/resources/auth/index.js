import axios from 'axios/index';
import dotenvConfiguration from '../../dotenv-configuration';

const AuthResource = {
  login(magicLinkToken) {
    return axios.get(`${dotenvConfiguration.API_URL}/auth/${magicLinkToken}`);
  },
};

export default AuthResource;
