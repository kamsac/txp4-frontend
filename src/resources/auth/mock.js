export const ACCESS_TOKEN = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwiaWF0IjoxNTE2MjM5MDIyLCJsb2dpbiI6Imx5bWFrIiwibmljayI6IiRzJG8kYTYwbCRjODJ5JGZhM20kYzgyYSRhNjBrIn0.aaKzXEZP2tovhBAqPJYBdmbLzKjfdBsuX3FaB58JZvE';

const AuthResourceMock = {
  login() {
    return new Promise((resolve) => {
      const response = {
        data: {
          token: ACCESS_TOKEN,
        },
      };
      setTimeout(() => resolve(response), 100);
    });
  },
};

export default AuthResourceMock;
