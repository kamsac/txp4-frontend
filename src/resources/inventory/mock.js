const ITEMS = [
  {
    id: "a1a1a1",
    name: "Tire",
    vendorId: 'aaa',
    equipRegion: "tire",
    tier: 3,
    isEquipped: true,
    modifiers: [
      {
        name: "maxSpeed",
        value: 1,
      },
      {
        name: "acceleration",
        value: 1.2,
      },
      {
        name: "steering",
        value: 1.1,
      },
      {
        name: "breaking",
        value: 1.2,
      },
      {
        name: "grip",
        value: 1.5,
      },
      {
        name: "gravity",
        value: 1,
      },
    ],
  },
  {
    id: "b2b2b2",
    name: "Less tired",
    vendorId: 'bbb',
    equipRegion: "tire",
    tier: 1,
    isEquipped: false,
    modifiers: [
      {
        name: "maxSpeed",
        value: 0.4,
      },
      {
        name: "acceleration",
        value: 1,
      },
      {
        name: "steering",
        value: 1.4,
      },
      {
        name: "breaking",
        value: 2,
      },
      {
        name: "grip",
        value: 3,
      },
      {
        name: "gravity",
        value: 1,
      },
    ],
  },
  {
    id: "c3c3c3",
    name: "Shit engine",
    vendorId: 'ccc',
    equipRegion: "engine",
    tier: 2,
    isEquipped: false,
    modifiers: [
      {
        name: "maxSpeed",
        value: 2,
      },
      {
        name: "acceleration",
        value: 1.8,
      },
      {
        name: "steering",
        value: 1,
      },
      {
        name: "breaking",
        value: 1,
      },
      {
        name: "grip",
        value: 0.2,
      },
      {
        name: "gravity",
        value: 1.5,
      },
    ],
  },
  {
    id: "d4d4d4",
    name: "Some tire",
    vendorId: 'ddd',
    equipRegion: "tire",
    tier: 2,
    isEquipped: false,
    modifiers: [
      {
        name: "maxSpeed",
        value: 1,
      },
      {
        name: "acceleration",
        value: 1,
      },
      {
        name: "steering",
        value: 1,
      },
      {
        name: "breaking",
        value: 1.1,
      },
      {
        name: "grip",
        value: 0.8,
      },
      {
        name: "gravity",
        value: 0.9,
      },
    ],
  },
  {
    id: "e5e5e5",
    name: "Tire",
    vendorId: 'aaa',
    equipRegion: "tire",
    tier: 3,
    isEquipped: false,
    modifiers: [
      {
        name: "maxSpeed",
        value: 1,
      },
      {
        name: "acceleration",
        value: 1.2,
      },
      {
        name: "steering",
        value: 1.1,
      },
      {
        name: "breaking",
        value: 1.2,
      },
      {
        name: "grip",
        value: 1.5,
      },
      {
        name: "gravity",
        value: 1,
      },
    ],
  },
  {
    id: "f6f6f6",
    name: "Duper engine",
    vendorId: 'aaa',
    equipRegion: "engine",
    tier: 2,
    isEquipped: false,
    modifiers: [
      {
        name: "maxSpeed",
        value: 1.5,
      },
      {
        name: "acceleration",
        value: 3,
      },
      {
        name: "steering",
        value: 1,
      },
      {
        name: "breaking",
        value: 1,
      },
      {
        name: "grip",
        value: 1,
      },
      {
        name: "gravity",
        value: 1.5,
      },
    ],
  },
  {
    id: "f7f7f7",
    name: "Tire",
    vendorId: 'aaa',
    equipRegion: "tire",
    tier: 3,
    isEquipped: false,
    modifiers: [
      {
        name: "maxSpeed",
        value: 1,
      },
      {
        name: "acceleration",
        value: 1.2,
      },
      {
        name: "steering",
        value: 1.1,
      },
      {
        name: "breaking",
        value: 1.2,
      },
      {
        name: "grip",
        value: 1.5,
      },
      {
        name: "gravity",
        value: 1,
      },
    ],
  },
  {
    id: "g8g8g8",
    name: "Less tired",
    vendorId: 'bbb',
    equipRegion: "tire",
    tier: 1,
    isEquipped: false,
    modifiers: [
      {
        name: "maxSpeed",
        value: 0.4,
      },
      {
        name: "acceleration",
        value: 1,
      },
      {
        name: "steering",
        value: 1.4,
      },
      {
        name: "breaking",
        value: 2,
      },
      {
        name: "grip",
        value: 3,
      },
      {
        name: "gravity",
        value: 1,
      },
    ],
  },
  {
    id: "h9h9h9",
    name: "Shit engine",
    vendorId: 'ccc',
    equipRegion: "engine",
    tier: 2,
    isEquipped: false,
    modifiers: [
      {
        name: "maxSpeed",
        value: 2,
      },
      {
        name: "acceleration",
        value: 1.8,
      },
      {
        name: "steering",
        value: 1,
      },
      {
        name: "breaking",
        value: 1,
      },
      {
        name: "grip",
        value: 0.2,
      },
      {
        name: "gravity",
        value: 1.5,
      },
    ],
  },
  {
    id: "i0i0i0",
    name: "Some tire",
    vendorId: 'ddd',
    equipRegion: "tire",
    tier: 2,
    isEquipped: false,
    modifiers: [
      {
        name: "maxSpeed",
        value: 1,
      },
      {
        name: "acceleration",
        value: 1,
      },
      {
        name: "steering",
        value: 1,
      },
      {
        name: "breaking",
        value: 1.1,
      },
      {
        name: "grip",
        value: 0.8,
      },
      {
        name: "gravity",
        value: 0.9,
      },
    ],
  },
  {
    id: "j1j1j1",
    name: "Transmission",
    vendorId: 'ccc',
    equipRegion: "transmission",
    tier: 3,
    isEquipped: false,
    modifiers: [
      {
        name: "maxSpeed",
        value: 0.7,
      },
      {
        name: "acceleration",
        value: 1.5,
      },
      {
        name: "steering",
        value: 1,
      },
      {
        name: "breaking",
        value: 1,
      },
      {
        name: "grip",
        value: 1,
      },
      {
        name: "gravity",
        value: 1,
      },
    ],
  },
];

const InventoryResourceMock = {
  getItems() {
    return new Promise((resolve, reject) => {
      const response = {
        data: ITEMS,
      };
      setTimeout(() => resolve(response), 100);
    });
  },

  equipItem(item) {
    return new Promise((resolve, reject) => {
      const response = {
        data: ITEMS,
      };
      setTimeout(() => resolve(ITEMS), 100);
    });
  },
};

export default InventoryResourceMock;
