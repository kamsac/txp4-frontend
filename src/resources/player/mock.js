export const ITEMS = [
  {
    id: 'a1a1a1',
    name: 'Tire',
    vendorId: 'aaa',
    equipRegion: 'tire',
    tier: 3,
    isEquipped: true,
    modifiers: [
      {
        name: 'maxSpeed',
        value: 1,
      },
      {
        name: 'acceleration',
        value: 1.2,
      },
      {
        name: 'steering',
        value: 1.1,
      },
      {
        name: 'breaking',
        value: 1.2,
      },
      {
        name: 'grip',
        value: 1.5,
      },
      {
        name: 'gravity',
        value: 1,
      },
    ],
  },
  {
    id: 'b2b2b2',
    name: 'Less tired',
    vendorId: 'bbb',
    equipRegion: 'tire',
    tier: 1,
    isEquipped: false,
    modifiers: [
      {
        name: 'maxSpeed',
        value: 0.4,
      },
      {
        name: 'acceleration',
        value: 1,
      },
      {
        name: 'steering',
        value: 1.4,
      },
      {
        name: 'breaking',
        value: 2,
      },
      {
        name: 'grip',
        value: 3,
      },
      {
        name: 'gravity',
        value: 1,
      },
    ],
  },
  {
    id: 'c3c3c3',
    name: 'Shit engine',
    vendorId: 'ccc',
    equipRegion: 'engine',
    tier: 2,
    isEquipped: false,
    modifiers: [
      {
        name: 'maxSpeed',
        value: 2,
      },
      {
        name: 'acceleration',
        value: 1.8,
      },
      {
        name: 'steering',
        value: 1,
      },
      {
        name: 'breaking',
        value: 1,
      },
      {
        name: 'grip',
        value: 0.2,
      },
      {
        name: 'gravity',
        value: 1.5,
      },
    ],
  },
  {
    id: 'd4d4d4',
    name: 'Some tire',
    vendorId: 'ddd',
    equipRegion: 'tire',
    tier: 2,
    isEquipped: false,
    modifiers: [
      {
        name: 'maxSpeed',
        value: 1,
      },
      {
        name: 'acceleration',
        value: 1,
      },
      {
        name: 'steering',
        value: 1,
      },
      {
        name: 'breaking',
        value: 1.1,
      },
      {
        name: 'grip',
        value: 0.8,
      },
      {
        name: 'gravity',
        value: 0.9,
      },
    ],
  },
  {
    id: 'e5e5e5',
    name: 'Tire',
    vendorId: 'aaa',
    equipRegion: 'tire',
    tier: 3,
    isEquipped: false,
    modifiers: [
      {
        name: 'maxSpeed',
        value: 1,
      },
      {
        name: 'acceleration',
        value: 1.2,
      },
      {
        name: 'steering',
        value: 1.1,
      },
      {
        name: 'breaking',
        value: 1.2,
      },
      {
        name: 'grip',
        value: 1.5,
      },
      {
        name: 'gravity',
        value: 1,
      },
    ],
  },
  {
    id: 'f6f6f6',
    name: 'Duper engine',
    vendorId: 'aaa',
    equipRegion: 'engine',
    tier: 2,
    isEquipped: false,
    modifiers: [
      {
        name: 'maxSpeed',
        value: 1.5,
      },
      {
        name: 'acceleration',
        value: 3,
      },
      {
        name: 'steering',
        value: 1,
      },
      {
        name: 'breaking',
        value: 1,
      },
      {
        name: 'grip',
        value: 1,
      },
      {
        name: 'gravity',
        value: 1.5,
      },
    ],
  },
  {
    id: 'f7f7f7',
    name: 'Tire',
    vendorId: 'aaa',
    equipRegion: 'tire',
    tier: 3,
    isEquipped: false,
    modifiers: [
      {
        name: 'maxSpeed',
        value: 1,
      },
      {
        name: 'acceleration',
        value: 1.2,
      },
      {
        name: 'steering',
        value: 1.1,
      },
      {
        name: 'breaking',
        value: 1.2,
      },
      {
        name: 'grip',
        value: 1.5,
      },
      {
        name: 'gravity',
        value: 1,
      },
    ],
  },
  {
    id: 'g8g8g8',
    name: 'Less tired',
    vendorId: 'bbb',
    equipRegion: 'tire',
    tier: 1,
    isEquipped: false,
    modifiers: [
      {
        name: 'maxSpeed',
        value: 0.4,
      },
      {
        name: 'acceleration',
        value: 1,
      },
      {
        name: 'steering',
        value: 1.4,
      },
      {
        name: 'breaking',
        value: 2,
      },
      {
        name: 'grip',
        value: 3,
      },
      {
        name: 'gravity',
        value: 1,
      },
    ],
  },
  {
    id: 'h9h9h9',
    name: 'Shit engine',
    vendorId: 'ccc',
    equipRegion: 'engine',
    tier: 2,
    isEquipped: false,
    modifiers: [
      {
        name: 'maxSpeed',
        value: 2,
      },
      {
        name: 'acceleration',
        value: 1.8,
      },
      {
        name: 'steering',
        value: 1,
      },
      {
        name: 'breaking',
        value: 1,
      },
      {
        name: 'grip',
        value: 0.2,
      },
      {
        name: 'gravity',
        value: 1.5,
      },
    ],
  },
  {
    id: 'i0i0i0',
    name: 'Some tire',
    vendorId: 'ddd',
    equipRegion: 'tire',
    tier: 2,
    isEquipped: false,
    modifiers: [
      {
        name: 'maxSpeed',
        value: 1,
      },
      {
        name: 'acceleration',
        value: 1,
      },
      {
        name: 'steering',
        value: 1,
      },
      {
        name: 'breaking',
        value: 1.1,
      },
      {
        name: 'grip',
        value: 0.8,
      },
      {
        name: 'gravity',
        value: 0.9,
      },
    ],
  },
  {
    id: 'j1j1j1',
    name: 'Transmission',
    vendorId: 'ccc',
    equipRegion: 'transmission',
    tier: 3,
    isEquipped: false,
    modifiers: [
      {
        name: 'maxSpeed',
        value: 0.7,
      },
      {
        name: 'acceleration',
        value: 1.5,
      },
      {
        name: 'steering',
        value: 1,
      },
      {
        name: 'breaking',
        value: 1,
      },
      {
        name: 'grip',
        value: 1,
      },
      {
        name: 'gravity',
        value: 1,
      },
    ],
  },
];

export const PLAYER = {
  login: 'lymak',
  nick: '$s$o$a60l$c82y$fa3m$c82a$a60k',
  inventory: {
    items: ITEMS,
  },
};

const PlayerResourceMock = {
  getPlayer() {
    return new Promise((resolve) => {
      const response = {
        data: PLAYER,
      };
      setTimeout(() => resolve(response), 100);
    });
  },

  getInventory() {
    return new Promise((resolve) => {
      const response = {
        data: {
          items: ITEMS,
        },
      };
      setTimeout(() => resolve(response), 100);
    });
  },

  equipItem() {
    return new Promise((resolve) => {
      const response = {
        data: {
          items: ITEMS,
        },
      };
      setTimeout(() => resolve(response), 100);
    });
  },

  unequipItem() {
    return new Promise((resolve) => {
      const response = {
        data: ITEMS,
      };
      setTimeout(() => resolve(response), 100);
    });
  },
};

export default PlayerResourceMock;
