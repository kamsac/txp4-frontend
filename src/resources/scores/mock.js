const SCORES = [
  {
    player: {
      login: 'lymak',
      nickname: 'lymak',
    },
    points: 4242,
  },
  {
    player: {
      login: 'kibes',
      nickname: 'kibes',
    },
    points: 3232,
  },
  {
    player: {
      login: 'tomek',
      nickname: 'tomek',
    },
    points: 2416,
  },
  {
    player: {
      login: 'omed',
      nickname: 'omed',
    },
    points: 2000,
  },
  {
    player: {
      login: 'tar',
      nickname: 'tar',
    },
    points: 1960,
  },
  {
    player: {
      login: 'oiz',
      nickname: 'oiz',
    },
    points: 680,
  },
  {
    player: {
      login: 'xetrow',
      nickname: 'xetrow',
    },
    points: 403,
  },
  {
    player: {
      login: 'ginok',
      nickname: 'ginok',
    },
    points: 30,
  },
  {
    player: {
      login: 'llif',
      nickname: 'llif',
    },
    points: 20,
  },
  {
    player: {
      login: 'ida',
      nickname: 'ida',
    },
    points: 10,
  },
];

const ScoresResourceMock = {
  getScores() {
    return new Promise((resolve, reject) => {
      const response = {
        data: SCORES,
      };
      setTimeout(() => resolve(response), 100);
    });
  },
};

export default ScoresResourceMock;