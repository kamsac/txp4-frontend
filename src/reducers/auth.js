import jwtDecode from 'jwt-decode';
import { LOGIN_FAILURE, LOGIN_REQUEST, LOGIN_SUCCESS, LOGOUT_REQUEST, LOGOUT_SUCCESS } from '../actions/auth';

function getDefaultJwtDecoded() {
  if (!localStorage.getItem('accessToken')) {
    return null;
  }
  return jwtDecode(localStorage.getItem('accessToken'));
}

export default function authReducer(
  state = {
    isFetching: false,
    jwtDecoded: getDefaultJwtDecoded(),
  },
  action = {
    type: 'nothing',
  },
) {
  switch (action.type) {
    case LOGIN_REQUEST:
      return Object.assign({}, state, {
        isFetching: action.payload.isFetching,
      });
    case LOGIN_SUCCESS:
      return Object.assign({}, state, {
        isFetching: action.payload.isFetching,
        jwtDecoded: action.payload.jwtDecoded,
      });
    case LOGIN_FAILURE:
      return Object.assign({}, state, {
        isFetching: action.payload.isFetching,
        jwtDecoded: action.payload.jwtDecoded,
      });
    case LOGOUT_REQUEST:
      return Object.assign({}, state, {
      });
    case LOGOUT_SUCCESS:
      return Object.assign({}, state, {
        jwtDecoded: action.payload.jwtDecoded,
      });
    default:
      return state;
  }
}

export function getIsUserAuthenticated(state) {
  return !!state.auth.jwtDecoded && !!state.auth.jwtDecoded.login;
}

export function getUserLogin(state) {
  return state.auth.jwtDecoded ? state.auth.jwtDecoded.login : null;
}

export function getUserNick(state) {
  return state.auth.jwtDecoded ? state.auth.jwtDecoded.nick : null;
}

export function getIsFetchingLogin(state) {
  return state.auth.isFetching;
}
