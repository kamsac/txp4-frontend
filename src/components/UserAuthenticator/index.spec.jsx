import React from 'react';
import { shallow } from 'enzyme';
import UserAuthenticator from './index';
import { ACCESS_TOKEN } from '../../resources/auth/mock';

describe('<UserAuthenticator>', () => {
  it('should render', () => {
    const component = shallow(<UserAuthenticator
      loginUser={() => {}}
      isFetching
      isUserAuthenticated={false}
      magicLinkToken={ACCESS_TOKEN}
    />);
    expect(component).toMatchSnapshot();
  });

  it('should try to authenticate', () => {
    const loginUserSpy = jest.fn();
    shallow(<UserAuthenticator
      loginUser={loginUserSpy}
      isFetching={false}
      isUserAuthenticated={false}
      magicLinkToken={ACCESS_TOKEN}
    />);
    expect(loginUserSpy).toBeCalledWith(ACCESS_TOKEN);
  });

  it('should redirect when authenticated to /inventory page', () => {
    const component = shallow(<UserAuthenticator
      loginUser={() => {}}
      isFetching={false}
      isUserAuthenticated
      magicLinkToken={ACCESS_TOKEN}
    />);
    const redirect = component.find('Redirect[to="/inventory"]');
    expect(redirect.length).toEqual(1);
  });
});
