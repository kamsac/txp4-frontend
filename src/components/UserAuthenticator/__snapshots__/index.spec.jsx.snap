// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserAuthenticator> should render 1`] = `
<p>
  You are being authenticated...
</p>
`;
